### ISSUE TITLE
[Error] Multiple non-persisted new entities were found through the given association graph

### ISSUE THREAD

**Issue by gnat42** [theofidry/AliceDataFixtures#95]
I'm stuck with trying to get some fixtures to load. Everything works until I add one set of entities at which point I get this error about new " A new entity was found through the relationship". I can build the same graph via a simple ContainerAwareCommand and it persists normally so I have no idea what's going on here.

Here are the fixtures:

```yml
# PracticeAndUsers.yml
---
NS\CoreDomain\Practice\Practice:
  unverified:
    __construct: [ '<company()>' ]
    street: '<streetAddress()>'
    city: '<city()>'
    province: '<state()>'
    postalCode: '<postcode()>'
    accountType: 'silver'
    verified: false
  practices{1..10}:
    __construct: [ '<company()>' ]
    street: '<streetAddress()>'
    city: '<city()>'
    province: "<state()>"
    postalCode: "<postcode()>"
    accountType: "<accountType()>"
    verified: true
    twilioNumber: '50%? <e164PhoneNumber()>'

NS\CoreDomain\Practice\User:
  unverifiedUser:
    __construct: [ "@unverified" ]
    name: '<name()>'
    email (unique): '<email()>'
    plainPassword: 1234567
  users{0..12}:
    __construct: [ "@practices*" ]
    name: '<name()>'
    email (unique): '<email()>'
    plainPassword: 1234567

# Clients.yml
---
NS\CoreDomain\Practice\AddressableEntity:
  addressableEntity{0..10}:
    __factory: { createPerson: [ '<lastName()>', '<firstName()>' ]}
  addressableEntity{11..30}:
    __construct: {createCorporation: [ '<company()>' ]}

NS\CoreDomain\Practice\Client:
  client{0..30}:
    __construct: [ '@practices*', '@addressableEntity*' ]

# MattersAndContracts.yml
---
NS\CoreDomain\Practice\Matter:
  realestate_matters{0..10}:
    __construct:
      0: ['@client*']
      1 (unique): '<bothify("RE-????-####")>'
      2: 'ns.realestate'
      3: '@users*'

# Everything loads fine if these are commented out
NS\CoreDomain\RealEstate\Contract:
  realestate_contracts{0..10}:
    __construct:
      0 (unique): '@realestate_matters*'
      1: 'purchase'
      2: false
```

Here's a snippet of command code that successfully creates the object I need without any error.

```php
        $practice = new Practice('Practice');
        $user = new User($practice);
        $user->setName('User Name');
        $user->setEmail($input->getArgument('email'));

        $person = AddressableEntity::createPerson('Noblet', 'Nathanael');
        $client = new Client($practice, $person);

        $matter = new Matter([$client], $input->getArgument('number'), MatterProvider::TYPE, $user);
        $contract = new Contract($matter, $input->getArgument('type'), false);

        $entityManager = $this->getContainer()->get('doctrine.orm.entity_manager');
        $entityManager->persist($user);
        $entityManager->persist($client);
        $entityManager->persist($practice);
        $entityManager->persist($matter);
        $entityManager->persist($contract);
        $entityManager->flush();
```
I've tried a few different things like changing the persisting order without any luck. Like I said I can do it manually so I'm confused what's going on.

**Comment by theofidry** [comment]
I would first try to:

- reduce the number of figures in order to identify the problematic relationship
- then try to identifier which properties being set causes the issue
- dump the graph of objects with the [Symfony VarDumper](https://symfony.com/doc/current/components/var_dumper.html) in order to check if the references of the objects are the ones you are expecting

Then it really depends on what's going on, it can be an issue in alice, your setters not being set like you expected, a clone somewhere in your code...

Also just double check if you are on the latests versions as well :)

Sorry if I cannot help more than that for now but it's really about your code until we have a clear reproducer

**Comment by gnat42** [comment]
Hello @theofidry 

* I checked my source code and nothing is cloned
* I made it so that there's only one realestate_contract but get the same issue
* I have hautelook/alice-bundle: v2.0.0 and theofidry/alice-data-fixtures: v1.0.1 which seem to be the latest
* I'll try to reduce everything to a simple set of objects to determine whats happening

Thank you for responding so quickly. Its a long weekend here so I won't be looking at this until probably Tuesday

**Comment by gnat42** [comment]
So I created a single yml file with a single object of each type

```yml
## YAML Template.
---
NS\CoreDomain\Practice\Practice:
  practice:
    __construct: [ '<company()>' ]
    street: '<streetAddress()>'
    city: '<city()>'
    province: "<state()>"
    postalCode: "<postcode()>"
    accountType: "<accountType()>"
    verified: true
    twilioNumber: '50%? <e164PhoneNumber()>'

NS\CoreDomain\Practice\User:
  user:
    __construct: [ "@practice" ]
    name: '<name()>'
    email (unique): '<email()>'
    plainPassword: 1234567

NS\CoreDomain\Practice\AddressableEntity:
  addressableEntity:
    __factory: { createPerson: [ '<lastName()>', '<firstName()>' ]}

NS\CoreDomain\Practice\Client:
  client:
    __construct: [ '@practice', '@addressableEntity' ]

NS\CoreDomain\Practice\Matter:
  realestate_matter:
    __construct:
      0: ['@client']
      1 (unique): '<bothify("RE-????-####")>'
      2: 'ns.realestate'
      3: '@user'

NS\CoreDomain\RealEstate\Contract:
  realestate_contract:
    __construct:
      0 (unique): '@realestate_matter'
      1: 'purchase'
      2: false
```
If the realestae_contract is not commented, I get the same error message. If I comment it, I get one of each object as expected.

They are the same objects too. For example, with the above I get the following error message:

```
 * A new entity was found through the relationship 'NS\CoreDomain\Practice\Client#practice' that was not configured to cascade persist operations for entity: Adams-Reichel....
 * A new entity was found through the relationship 'NS\CoreDomain\Practice\Matter#lawyer' that was not configured to cascade persist operations for entity: Dr. Samir Weimann DDS....
```
The name of the practice is Adams-Reichel which is the name the practice that gets successfully persisted when the realestate_contracts is commented out. Same goes for the lawyer.

**Comment by gnat42** [comment]
This is the output of symfony's dump of the client object and the contract object. You can see that the client object '#2549 isn't the same as the client in the matter object '#2713 which is copied into the contract. I don't get why. I've included the dump and the Matter constructor.

```
NS\CoreDomain\Practice\Client {#2549
  -id: null
  -entity: NS\CoreDomain\Practice\AddressableEntity {#2545
    -lastName: "Thiel"
    -givenName: "Wanda"
  }
  -practice: NS\CoreDomain\Practice\Practice {#2465
    -id: null
    -name: "Adams-Reichel"
    -hash: "8bff1f8f618444cfe7e6d965e611dae5faa4f809"
    -clients: Doctrine\Common\Collections\ArrayCollection {#2519
      -elements: []
    }
  }
}


NS\CoreDomain\RealEstate\Contract {#2671
  -id: null
  -matterNumber: "RE-ixko-4622"
  -matter: NS\CoreDomain\Practice\Matter {#2658
    -id: null
    -practice: NS\CoreDomain\Practice\Practice {#2712
      -id: null
      -name: "Adams-Reichel"
      -hash: "8bff1f8f618444cfe7e6d965e611dae5faa4f809"
    }
    -clients: array:1 [
      0 => NS\CoreDomain\Practice\Client {#2713
        -id: null
        -entity: NS\CoreDomain\Practice\AddressableEntity {#2729
          -lastName: "Thiel"
          -givenName: "Wanda"
        }
        -practice: NS\CoreDomain\Practice\Practice {#2712}
      }
    ]
    -number: "RE-ixko-4622"
    -target: "ns.realestate-purchase"
    -lawyer: NS\CoreDomain\Practice\User {#2737
      -id: null
      -name: "Dr. Samir Weimann DDS"
      -email: "rau.celia@parker.com"
      -type: "lawyer"
      -practice: NS\CoreDomain\Practice\Practice {#2712}
    }
  }
  -status: "Open"
  -type: "purchase"
  -clients: array:1 [
    0 => NS\CoreDomain\Practice\Client {#2713}
  ]
  -practice: NS\CoreDomain\Practice\Practice {#2712}
}
```

```php

    public function __construct(array $clients, string $number, string $target, User $lawyer, ?User $assistant = null)
    {
        if ($lawyer->getType() !== 'lawyer') {
            throw new \InvalidArgumentException('Invalid assigned user. Expecting user of type lawyer');
        }

        $lastPractice = null;
        foreach ($clients as $client) {
            if ($client === null) {
                continue;
            }

            if (!$client instanceof Client) {
                throw new \InvalidArgumentException("Not a client ".gettype($client).' '.get_class($client));
            }

            if ($lastPractice === null) {
                $lastPractice = $client->getPractice();
            }

            if ($lastPractice !== $client->getPractice()) {
                throw new \InvalidArgumentException();
            }
        }

        if ($lastPractice === null) {
            throw new \InvalidArgumentException("No practice so no clients?");
        }

        $this->practice       = $lastPractice;
        $this->clients        = $clients;
        $this->target         = $target;
        $this->lawyer         = $lawyer;
        $this->number         = $number;
        $this->communications = new ArrayCollection();
        $this->createdOn      = new \DateTime();
        $this->workSteps      = new ArrayCollection();
        $this->assistant      = $assistant;
    }
```

**Comment by gnat42** [comment]
Soooo I found the issue, but don't know what the solution is. Since none of my code uses clone, I started looking at the bundles related to alice. I added __clone to Client with an exception and the stack trace led me to:

```
NS\CoreDomain\Practice\Client->__clone() at /var/www/local/prolegis/vendor/myclabs/deep-copy/src/DeepCopy/DeepCopy.php:185
 DeepCopy\DeepCopy->copyObject() at /var/www/local/prolegis/vendor/myclabs/deep-copy/src/DeepCopy/DeepCopy.php:133
 DeepCopy\DeepCopy->recursiveCopy() at /var/www/local/prolegis/vendor/myclabs/deep-copy/src/DeepCopy/DeepCopy.php:144
 DeepCopy\DeepCopy->copyArray() at /var/www/local/prolegis/vendor/myclabs/deep-copy/src/DeepCopy/DeepCopy.php:124
 DeepCopy\DeepCopy->recursiveCopy() at /var/www/local/prolegis/vendor/myclabs/deep-copy/src/DeepCopy/DeepCopy.php:235
 DeepCopy\DeepCopy->copyObjectProperty() at /var/www/local/prolegis/vendor/myclabs/deep-copy/src/DeepCopy/DeepCopy.php:197
 DeepCopy\DeepCopy->copyObject() at /var/www/local/prolegis/vendor/myclabs/deep-copy/src/DeepCopy/DeepCopy.php:133
 DeepCopy\DeepCopy->recursiveCopy() at /var/www/local/prolegis/vendor/myclabs/deep-copy/src/DeepCopy/DeepCopy.php:91
 DeepCopy\DeepCopy->copy() at /var/www/local/prolegis/vendor/nelmio/alice/src/deep_clone.php:28
 Nelmio\Alice\deep_clone() at /var/www/local/prolegis/vendor/nelmio/alice/src/Definition/Value/UniqueValue.php:44
 Nelmio\Alice\Definition\Value\UniqueValue->__construct() at /var/www/local/prolegis/vendor/nelmio/alice/src/Definition/Value/UniqueValue.php:49
```

Which pointed to the (unique) constraint I have on the realestate_contracts constructor. This changes my understanding of the (unique) clause on the fixture files. I thought it would simply make sure it never re-used one of the references, but this seems to clone each one.

So, other than providing my own provider function that gets all references to objects and never re-uses one, what's the solution?

**Comment by theofidry** [comment]
I'm a bit surprised by that result because the `Definition\Value\*` shouldn't hold any references to the existing values, i.e. when you have:

```
Acme\Dummy:
  dummy:
    foo (unique): '@bar'
```

You get something like `UniqueValue(New FixtureReference('bar'))`. There may be some exceptions and `UniqueValue` might be one of them. It's worth checking why is that and if the clone can be removed there

**Comment by gnat42** [comment]
Yeah, when I change

```
NS\CoreDomain\RealEstate\Contract:
  realestate_contract:
    __construct:
      0 (unique): '@realestate_matter'
      1: 'purchase'
      2: false
```

To

```
NS\CoreDomain\RealEstate\Contract:
  realestate_contract:
    __construct:
      0: '@realestate_matter<current()>'
      1: 'purchase'
      2: false
```
Everything works fine.

UniqueValue's constructor has no logic beside doing a clone:

```
    public function __construct(string $id, $value)
    {
        $this->id = $id;

        if ($value instanceof self) {
            throw InvalidArgumentExceptionFactory::createForRedundantUniqueValue($id);
        }

        $this->value = deep_clone($value);
    }
```

**Comment by gnat42** [comment]
``` NS\CoreDomain\Practice\Client->__clone() at /var/www/local/prolegis/vendor/myclabs/deep-copy/src/DeepCopy/DeepCopy.php:185
 DeepCopy\DeepCopy->copyObject() at /var/www/local/prolegis/vendor/myclabs/deep-copy/src/DeepCopy/DeepCopy.php:133
 DeepCopy\DeepCopy->recursiveCopy() at /var/www/local/prolegis/vendor/myclabs/deep-copy/src/DeepCopy/DeepCopy.php:144
 DeepCopy\DeepCopy->copyArray() at /var/www/local/prolegis/vendor/myclabs/deep-copy/src/DeepCopy/DeepCopy.php:124
 DeepCopy\DeepCopy->recursiveCopy() at /var/www/local/prolegis/vendor/myclabs/deep-copy/src/DeepCopy/DeepCopy.php:235
 DeepCopy\DeepCopy->copyObjectProperty() at /var/www/local/prolegis/vendor/myclabs/deep-copy/src/DeepCopy/DeepCopy.php:197
 DeepCopy\DeepCopy->copyObject() at /var/www/local/prolegis/vendor/myclabs/deep-copy/src/DeepCopy/DeepCopy.php:133
 DeepCopy\DeepCopy->recursiveCopy() at /var/www/local/prolegis/vendor/myclabs/deep-copy/src/DeepCopy/DeepCopy.php:91
 DeepCopy\DeepCopy->copy() at /var/www/local/prolegis/vendor/nelmio/alice/src/deep_clone.php:28
 Nelmio\Alice\deep_clone() at /var/www/local/prolegis/vendor/nelmio/alice/src/Definition/Value/UniqueValue.php:44
 Nelmio\Alice\Definition\Value\UniqueValue->__construct() at /var/www/local/prolegis/vendor/nelmio/alice/src/Definition/Value/UniqueValue.php:49
 Nelmio\Alice\Definition\Value\UniqueValue->withValue() at /var/www/local/prolegis/vendor/nelmio/alice/src/Generator/Resolver/Value/Chainable/UniqueValueResolver.php:139
 Nelmio\Alice\Generator\Resolver\Value\Chainable\UniqueValueResolver->generateValue() at /var/www/local/prolegis/vendor/nelmio/alice/src/Generator/Resolver/Value/Chainable/UniqueValueResolver.php:100
 Nelmio\Alice\Generator\Resolver\Value\Chainable\UniqueValueResolver->resolve() at /var/www/local/prolegis/vendor/nelmio/alice/src/Generator/Resolver/Value/ValueResolverRegistry.php:84
 Nelmio\Alice\Generator\Resolver\Value\ValueResolverRegistry->resolve() at /var/www/local/prolegis/vendor/nelmio/alice/src/Generator/Instantiator/InstantiatorResolver.php:139
 Nelmio\Alice\Generator\Instantiator\InstantiatorResolver->resolveArguments() at /var/www/local/prolegis/vendor/nelmio/alice/src/Generator/Instantiator/InstantiatorResolver.php:107
 Nelmio\Alice\Generator\Instantiator\InstantiatorResolver->resolveFixtureConstructor() at /var/www/local/prolegis/vendor/nelmio/alice/src/Generator/Instantiator/InstantiatorResolver.php:78
 Nelmio\Alice\Generator\Instantiator\InstantiatorResolver->instantiate() at /var/www/local/prolegis/vendor/nelmio/alice/src/Generator/Instantiator/ExistingInstanceInstantiator.php:66
 Nelmio\Alice\Generator\Instantiator\ExistingInstanceInstantiator->instantiate() at /var/www/local/prolegis/vendor/nelmio/alice/src/Generator/ObjectGenerator/SimpleObjectGenerator.php:85
 Nelmio\Alice\Generator\ObjectGenerator\SimpleObjectGenerator->generate() at /var/www/local/prolegis/vendor/nelmio/alice/src/Generator/ObjectGenerator/CompleteObjectGenerator.php:53
 Nelmio\Alice\Generator\ObjectGenerator\CompleteObjectGenerator->generate() at /var/www/local/prolegis/vendor/nelmio/alice/src/Generator/DoublePassGenerator.php:61
 Nelmio\Alice\Generator\DoublePassGenerator->generateFixtures() at /var/www/local/prolegis/vendor/nelmio/alice/src/Generator/DoublePassGenerator.php:49
 Nelmio\Alice\Generator\DoublePassGenerator->generate() at /var/www/local/prolegis/vendor/nelmio/alice/src/Loader/SimpleDataLoader.php:49
 Nelmio\Alice\Loader\SimpleDataLoader->loadData() at /var/www/local/prolegis/vendor/nelmio/alice/src/Loader/SimpleFilesLoader.php:65
 Nelmio\Alice\Loader\SimpleFilesLoader->loadFiles() at /var/www/local/prolegis/vendor/theofidry/alice-data-fixtures/src/Loader/SimpleLoader.php:53
 Fidry\AliceDataFixtures\Loader\SimpleLoader->load() at /var/www/local/prolegis/var/cache/test/ContainerIxehhag/SimpleLoader_4473cb1.php:27
 SimpleLoader_4473cb1->load() at /var/www/local/prolegis/vendor/theofidry/alice-data-fixtures/src/Loader/PersisterLoader.php:77
 Fidry\AliceDataFixtures\Loader\PersisterLoader->load() at /var/www/local/prolegis/vendor/theofidry/alice-data-fixtures/src/Loader/PurgerLoader.php:119
 Fidry\AliceDataFixtures\Loader\PurgerLoader->load() at /var/www/local/prolegis/vendor/theofidry/alice-data-fixtures/src/Loader/FileResolverLoader.php:73
 Fidry\AliceDataFixtures\Loader\FileResolverLoader->load() at /var/www/local/prolegis/vendor/hautelook/alice-bundle/src/Loader/DoctrineOrmLoader.php:184
 Hautelook\AliceBundle\Loader\DoctrineOrmLoader->loadFixtures() at /var/www/local/prolegis/vendor/hautelook/alice-bundle/src/Loader/DoctrineOrmLoader.php:117
 Hautelook\AliceBundle\Loader\DoctrineOrmLoader->load() at /var/www/local/prolegis/vendor/hautelook/alice-bundle/src/Console/Command/Doctrine/DoctrineOrmLoadDataFixturesCommand.php:138
 Hautelook\AliceBundle\Console\Command\Doctrine\DoctrineOrmLoadDataFixturesCommand->execute() at /var/www/local/prolegis/vendor/symfony/symfony/src/Symfony/Component/Console/Command/Command.php:251
 Symfony\Component\Console\Command\Command->run() at /var/www/local/prolegis/vendor/symfony/symfony/src/Symfony/Component/Console/Application.php:964
```

**Comment by gnat42** [comment]
Anything else I should do to help debug this issue?

**Comment by theofidry** [comment]
IIRC there was
https://github.com/nelmio/alice/pull/938 trying to fix it, maybe you can
try that branch and see if that fixes your issue

**Comment by theofidry** [comment]
Closing as it should have been fixed by #938. If that's not the case, please tell me so I'll open the issue again
